lib/ecto/embedded.ex

```diff
@@ -87,12 +101,20 @@ defmodule Ecto.Embedded do
   end
 
   defp apply_callback(%Changeset{model: %{__struct__: model}, action: action} = changeset,
-                      model, embed, adapter, function, type) do
+                      model, embed, adapter, function, :before) do
     check_action!(action, function, model)
-    callback = callback_for(type, action)
+    callback = callback_for(:before, action)
     Ecto.Model.Callbacks.__apply__(model, callback, changeset)
     |> generate_id(callback, model, embed, adapter)
-    |> apply_callbacks(model.__schema__(:embeds), adapter, function, type)
+    |> apply_callbacks(model.__schema__(:embeds), adapter, function, :before)
+  end
+
+  defp apply_callback(%Changeset{model: %{__struct__: model}, action: action} = changeset,
```

Comment:
No, because the order of running callbacks was wrong. We want to run parent callbacks first on before, and child callbacks first on after. It was wrong before with running all child callbacks after parents.